app/src/org/commcare/android/util/FormUploadUtil.java

```diff
@@ -63,255 +70,308 @@ public static Cipher getDecryptCipher(SecretKeySpec key) {
+            for (String ext : SUPPORTED_FILE_EXTS) {
+                if (f.getName().endsWith(ext)) {
                     supported = true;
                     break;
                 }
             }
-            
-            //TODO: Match this with some reasonable library, rather than silly file lines
+
             if (f.getName().endsWith(".xml")) {
-                
-                //fb = new InputStreamBody(new CipherInputStream(new FileInputStream(f), getDecryptCipher(aesKey)), "text/xml", f.getName());
-                
-                if(key != null){
-                    if(!validateSubmissionFile(f, FormUploadUtil.getDecryptCipher(key))) {
-                        return RECORD_FAILURE;
+                if (key != null) {
+                    if (!validateSubmissionFile(f)) {
+                        return false;
                     }
-                    fb = new EncryptedFileBody(f, FormUploadUtil.getDecryptCipher(key), "text/xml");
-                }
-                else{
+                    fb = new EncryptedFileBody(f, FormUploadUtil.getDecryptCipher(key),
+                            "text/xml");
+                } else {
                     fb = new FileBody(f, "text/xml");
                 }
-                
                 entity.addPart("xml_submission_file", fb);
-                
             } else if (f.getName().endsWith(".jpg")) {
                 fb = new FileBody(f, "image/jpeg");
                 if (fb.getContentLength() <= MAX_BYTES) {
                     entity.addPart(f.getName(), fb);
-                    Log.i(t, "added image file " + f.getName());
+                    Log.i(TAG, "added image file " + f.getName());
                 } else {
-                    Log.i(t, "file " + f.getName() + " is too big");
+                    Log.i(TAG, "file " + f.getName() + " is too big");
                 }
             } else if (f.getName().endsWith(".3gpp")) {
                 fb = new FileBody(f, "audio/3gpp");
                 if (fb.getContentLength() <= MAX_BYTES) {
                     entity.addPart(f.getName(), fb);
-                    Log.i(t, "added audio file " + f.getName());
+                    Log.i(TAG, "added audio file " + f.getName());
                 } else {
-                    Log.i(t, "file " + f.getName() + " is too big");
+                    Log.i(TAG, "file " + f.getName() + " is too big");
                 }
             } else if (f.getName().endsWith(".3gp")) {
                 fb = new FileBody(f, "video/3gpp");
                 if (fb.getContentLength() <= MAX_BYTES) {
                     entity.addPart(f.getName(), fb);
-                    Log.i(t, "added video file " + f.getName());
+                    Log.i(TAG, "added video file " + f.getName());
+                } else {
+                    Log.i(TAG, "file " + f.getName() + " is too big");
+                }
+            } else if (supported || isAudioVisualMimeType(f)) {
```

Comment:
The actual bug fix. Adds queued up files that have correct mimetype to the multipart entity to be submitted to HQ.